### PR TITLE
WIP: Add annotation to indicate parent workload by Kueue

### PR DESCRIPTION
**Pull request by mimowo**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

It is needed for the Kueue integration with other frameworks. The first integration we will need it for is with the MPI operator framework. The annotation will be added by a webhook to the child job created by a framework (the Launcher job in case of the MPIJob). A job with this annotation will not create a Kueue workload, but reuse the workload created by its parent job (the MPIJob in the case of the integration with MPIJob).

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

WIP, because not tested yet.

### CODE REVIEW

File: pkg/controller/workload/job/job_controller.go [hunk 1]
```diff
@@ -100,16 +105,76 @@ func NewReconciler(
...
+	if !ok {
+		return
+	}
+	ctx := context.TODO()
+	log := ctrl.LoggerFrom(ctx).WithValues("workload", klog.KObj(w))
+	ctx = ctrl.LoggerInto(ctx, log)
+	log.V(5).Info("Queueing reconcile for child jobs")
+	var childJobs batchv1.JobList
+	if err := h.client.List(ctx, &childJobs, client.InNamespace(w.Namespace), client.MatchingFields{parentWorkloadKey: w.Name}); err != nil {
+		klog.Error(err, "Unable to list child workloads")
```
Comment (tenzen-y) [inline pkg/controller/workload/job/job_controller.go:142]:
```suggestion
		klog.Error(err, "Unable to list child jobs")
```

**Review by tenzen-y — COMMENTED** [review]
@mimowo Looks great! Thanks for this!

File: pkg/controller/workload/job/job_controller.go [hunk 2]
```diff
@@ -100,16 +105,76 @@ func NewReconciler(
...
+// TODO: replace the TODO context with the one passed to the event handler's functions
+// when a new version of controller-runtime is used. See in master:
+// https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/handler/eventhandler.go
+func (h *jobWorkloadHandler) queueReconcileForChildJob(object client.Object, q workqueue.RateLimitingInterface) {
+	w, ok := object.(*kueue.Workload)
+	if !ok {
+		return
+	}
+	ctx := context.TODO()
+	log := ctrl.LoggerFrom(ctx).WithValues("workload", klog.KObj(w))
```
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/controller/workload/job/job_controller.go:137]: I'm not sure I can do that - I use the ctx variable to pass it into `LoggerInto`. This basically follows the pattern in the `Reconcile` function.

File: pkg/constants/constants.go [hunk 3]
```diff
@@ -24,6 +24,11 @@ const (
 	// TODO(#23): Use the kubernetes.io domain when graduating APIs to beta.
 	QueueAnnotation = "kueue.x-k8s.io/queue-name"
 
+	// ParentWorkloadAnnotation is the annotation used to mark child jobs to be
+	// controller by Kueue and ignored from scheduling. It's value is the name
+	// of the workload controller by the main job.
+	ParentWorkloadAnnotation = "kueue.x-k8s.io/parent-workload"
```
Comment (kerthcet) [inline pkg/constants/constants.go:30]: But what's the relationship with kubernetes batch Job? MPI job is migrating to use batch Job instead?

File: pkg/controller/workload/job/job_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/workload/job/job_controller.go:137]: Ah, I see. I'm fine with your implementation.

File: pkg/constants/constants.go (hunk 3 above)
Comment (tenzen-y) [reply pkg/constants/constants.go:30]: Yes, v2 creates the Launcher via batch/v1 Job.

Comment (tenzen-y) [reply pkg/constants/constants.go:30]: https://github.com/kubeflow/mpi-operator/blob/c21942d1e27d196d06078f67f39a6f7a0e990dee/pkg/controller/mpi_job_controller.go#L1390-L1413

File: pkg/constants/constants.go [hunk 4]
```diff
@@ -24,6 +24,11 @@ const (
 	// TODO(#23): Use the kubernetes.io domain when graduating APIs to beta.
 	QueueAnnotation = "kueue.x-k8s.io/queue-name"
 
+	// ParentWorkloadAnnotation is the annotation used to mark child jobs to be
+	// controller by Kueue and ignored from scheduling. It's value is the name
+	// of the workload controller by the main job.
```
Comment (alculquicondor) [inline pkg/constants/constants.go:29]: is the annotation used to mark a Job as a child of a Workload. This can be used when a custom job CRD is composed of one or more kubernetes Jobs. When set, Kueue ignores this Job from admission. The value is the name of the Workload that controls this Job, in the same namespace.

File: pkg/constants/constants.go (hunk 3 above)
Comment (alculquicondor) [reply pkg/constants/constants.go:30]: Can you validate in the webhook that the value is a valid CRD name?
Maybe we should also validate that it's immutable?

File: pkg/controller/workload/job/job_controller.go [hunk 5]
```diff
@@ -100,16 +105,76 @@ func NewReconciler(
 	}
 }
 
+type jobWorkloadHandler struct {
```
Comment (alculquicondor) [inline pkg/controller/workload/job/job_controller.go:108]: nit:

```suggestion
type parentWorkloadHandler struct {
```

I don't think we need `job` in the name as this is the job package.

File: pkg/constants/constants.go (hunk 4 above)
Comment (alculquicondor) [reply pkg/constants/constants.go:29]: However, it's not totally ignored. If the workload doesn't exist or it's not admitted, we should suspend the job.

File: pkg/controller/workload/job/job_controller.go [hunk 6]
```diff
@@ -173,6 +238,10 @@ func (r *JobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
 		if jobFinished {
 			return ctrl.Result{}, nil
 		}
+		if parentWorkload(&job) != "" {
+			log.V(3).Info("The parent workload '%s' does not exist", parentWorkload(&job))
```
Comment (alculquicondor) [inline pkg/controller/workload/job/job_controller.go:242]: Not sure if we should log anything here, or maybe we should give more information.

Is there already a log line explaining that the job was suspended? If it exists, maybe we should add a field for the missing Workload there. If not, we can say here that the Workload was suspended.

Also, we use structured logging, not `%s`.

File: pkg/controller/workload/job/job_controller.go [hunk 7]
```diff
@@ -173,6 +238,10 @@ func (r *JobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
...
+			return ctrl.Result{}, nil
+		}
 		err := r.handleJobWithNoWorkload(ctx, &job)
 		if err != nil {
 			log.Error(err, "Handling job with no workload")
```
Comment (alculquicondor) [inline pkg/controller/workload/job/job_controller.go:247]: 3. handle a finished job

should not be done for jobs with a parent workload.

File: pkg/constants/constants.go (hunk 3 above)
Comment (tenzen-y) [reply pkg/constants/constants.go:30]: > Can you validate in the webhook that the value is a valid CRD name?
Maybe we should also validate that it's immutable?

I agree.

File: pkg/controller/workload/job/job_controller.go (hunk 7 above)
Comment (mimowo) [reply pkg/controller/workload/job/job_controller.go:247]: Fixed

File: pkg/controller/workload/job/job_controller.go (hunk 5 above)
Comment (mimowo) [reply pkg/controller/workload/job/job_controller.go]: Done

File: pkg/constants/constants.go (hunk 4 above)
Comment (mimowo) [reply pkg/constants/constants.go]: rephrased, PTAL

File: pkg/constants/constants.go (hunk 3 above)
Comment (mimowo) [reply pkg/constants/constants.go:33]: Job webhook extended, including the check to ensure it is immutable.

File: pkg/controller/workload/job/job_controller.go (hunk 6 above)
Comment (mimowo) [reply pkg/controller/workload/job/job_controller.go]: This remark led me to identify a bug - a job without a parent workload wound't be suspended. There is an integration test aiming to test that, but it was starting from already suspended job. 

Now fixed the test and implementation. I also drop this log like, as I added another one to mention that the parent workload is missing just after trying to get it by the client.

File: pkg/controller/workload/job/job_controller.go [hunk 8]
```diff
@@ -373,6 +443,23 @@ func (r *JobReconciler) ensureAtMostOneWorkload(ctx context.Context, job *batchv
 	// Find a matching workload first if there is one.
 	var toDelete []*kueue.Workload
 	var match *kueue.Workload
+
+	if parentWorkload(job) != "" {
```
Comment (denkensk) [inline pkg/controller/workload/job/job_controller.go:447]: better to define `pwName := parentWorkload(job)` and use `pwName` in other places

Comment (mimowo) [reply pkg/controller/workload/job/job_controller.go]: Done

File: pkg/controller/workload/job/job_controller.go [hunk 9]
```diff
@@ -165,6 +230,7 @@ func (r *JobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
 		log.Error(err, "Getting existing workloads")
```
Comment (alculquicondor) [inline pkg/controller/workload/job/job_controller.go:230]: I think we need to adjust the line above that says:
```golang
if queueName(&job) == "" && !r.manageJobsWithoutQueueName
```

We still need to reconcile jobs that have the `parent-workload` annotation.

File: pkg/controller/workload/job/job_webhook.go [hunk 10]
```diff
@@ -100,6 +117,11 @@ func validateUpdate(oldJob, newJob *batchv1.Job) error {
 		return field.Forbidden(suspendPath, "should not update queue name when job is unsuspend")
 	}
 
+	oldValue, oldExists := oldJob.Annotations[constants.ParentWorkloadAnnotation]
+	newValue, newExists := newJob.Annotations[constants.ParentWorkloadAnnotation]
+	if oldValue != newValue || oldExists != newExists {
+		return field.Forbidden(parentWorkloadKeyPath, "this annotation is immutable")
```
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/controller/workload/job/job_webhook.go:122]: Done, also adjusted a test which was sensitive to nil vs. "" whereas this helper method is not

File: pkg/controller/workload/job/job_webhook_test.go [hunk 11]
```diff
@@ -27,6 +27,40 @@ import (
...
+			wantErr: field.Invalid(parentWorkloadKeyPath, "parent workload name", `a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')`),
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			gotErr := validateCreate(tc.job)
+
+			if diff := cmp.Diff(tc.wantErr, gotErr); diff != "" {
+				t.Errorf("validateUpdate() mismatch (-want +got):\n%s", diff)
```
Comment (tenzen-y) [inline pkg/controller/workload/job/job_webhook_test.go:58]:
```suggestion
				t.Errorf("validateCreate() mismatch (-want +got):\n%s", diff)
```

File: pkg/controller/workload/job/job_controller.go [hunk 12]
```diff
@@ -100,16 +105,76 @@ func NewReconciler(
...
 func (r *JobReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	wlHandler := parentWorkloadHandler{client: r.client}
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&batchv1.Job{}).
+		Watches(&source.Kind{Type: &kueue.Workload{}}, &wlHandler).
 		Owns(&kueue.Workload{}).
 		Complete(r)
 }
 
 func SetupIndexes(indexer client.FieldIndexer) error {
```
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/controller/workload/job/job_controller.go:167]: Can you explain which function concretely? As indicated in a comment, eventhandlers accept context, but only in master, not in the version we use. The Watches function does accept the context itself, and the context is also not passed currently via the wrapping method `SetupWithManager`.

Anyway, I guess I would anyway prefer to do it as a follow up.

File: pkg/controller/workload/job/job_webhook.go [hunk 13]
```diff
@@ -76,6 +83,16 @@ var _ webhook.CustomValidator = &JobWebhook{}
 
 // ValidateCreate implements webhook.CustomValidator so a webhook will be registered for the type
 func (w *JobWebhook) ValidateCreate(ctx context.Context, obj runtime.Object) error {
+	job := obj.(*batchv1.Job)
+	return validateCreate(job)
+}
+
+func validateCreate(job *batchv1.Job) error {
+	if value, exists := job.Annotations[constants.ParentWorkloadAnnotation]; exists {
```
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/controller/workload/job/job_webhook.go:92]: I prefer a follow up so that it is clear.

File: pkg/controller/workload/job/job_controller.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/controller/workload/job/job_controller.go:167]: `SetupIndexes`

File: pkg/controller/workload/job/job_controller.go (hunk 9 above)
Comment (alculquicondor) [reply pkg/controller/workload/job/job_controller.go:229]: This is a blocker for the PR

File: pkg/controller/workload/job/job_controller.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/workload/job/job_controller.go:167]: I guess we can propagate the following context.

https://github.com/kubernetes-sigs/kueue/blob/main/main.go#L126

[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/workload/job/job_controller.go:167]:
```diff
- func setupIndexes(mgr ctrl.Manager) {
+ func setupIndexes(ctx context.Context, mgr ctrl.Manager) {
...
-	if err := job.SetupIndexes(mgr.GetFieldIndexer()); err != nil {
+	if err := job.SetupIndexes(ctx, mgr.GetFieldIndexer()); err != nil {
		setupLog.Error(err, "Unable to setup job indexes")
	}
}
```

[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/controller/workload/job/job_controller.go:167]: Ok, and then use it in the eventhandlers? 

IIUC, the new event handlers in master would have passed by the controller-runtime framework, this means that we may then need to revert as the change then?

File: pkg/controller/workload/job/job_controller.go (hunk 9 above)
Comment (mimowo) [reply pkg/controller/workload/job/job_controller.go:232]: Done, this was actually a bug, that is not covered with an integration test that failed before the fix.

**Comment by mimowo** [comment]
Squashed the commits

File: pkg/controller/workload/job/job_webhook.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/controller/workload/job/job_webhook.go:92]: Opened: https://github.com/kubernetes-sigs/kueue/issues/562

File: pkg/controller/workload/job/job_controller.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/workload/job/job_controller.go:167]: > Ok, and then use it in the eventhandlers?

IIUC, we use it in the L168.

[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/controller/workload/job/job_controller.go:167]: `SetupIndexes` is not called by the event handlers. It's called during setup, which does have context.

**Review by alculquicondor — COMMENTED** [review]
/approve
Leaving the lgtm to @tenzen-y

**Review by tenzen-y — COMMENTED** [review]
Thanks!
/lgtm

File: pkg/controller/workload/job/job_controller.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/controller/workload/job/job_controller.go:167]: Opened https://github.com/kubernetes-sigs/kueue/issues/563.
